File: Design-Patterns/TemplateMethod/Cpp/spell.cpp
#include "spell.h"
#include <stdexcept>
#include <iostream>

float spell::get_cool_down_counter()
{
	return cool_down_counter_;
}

void spell::set_validator(std::shared_ptr<spell_casting_validator> validator)
{
	if (validator == nullptr)
	{
		throw std::invalid_argument("Validator is set to nullptr");
	}

	validator_ = validator;
}

void spell::cast()
{
	if (validator_ == nullptr || validator_->is_castable())
	{
		std::cout << "Casts spell" << std::endl;
	}
}

File: Design-Patterns/TemplateMethod/CSharp/Spell.cs
using System;
using TemplateMethod.SpellCastingValidators;

namespace TemplateMethod
{
    public class Spell
    {
        public float CoolDownCounter { get; protected set; }

        public SpellCastingValidator Validator
        {
            set
            {
                if (value == null)
                {
                    throw new ArgumentException();
                }
                else
                {
                    _validator = value;
                }
            }
        }

        private SpellCastingValidator _validator;

        public void Cast()
        {
            if (_validator == null || _validator.IsCastable())
            {
                Console.WriteLine("Casts spell");
            }
        }
    }
}
